### PR TITLE
UCT/IB/MLX5: Separate DV/EXP implementations

### PR DESCRIPTION
**Pull request by Artemy-Mellanox**

### CODE REVIEW

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/10620/ for details (Mellanox internal link).

**Comment by Artemy-Mellanox** [comment]
@yosefe

File: src/uct/ib/base/ib_md.h [hunk 1]
```diff
@@ -157,22 +157,28 @@ typedef struct uct_ib_rcache_region {
...
             .ops = &_md_ops, \
+            .priority = _priority, \
         }; \
-        if (_priority) { \
+        if (ucs_list_is_empty(&uct_ib_md_ops_list)) { \
             ucs_list_add_head(&uct_ib_md_ops_list, &entry.list); \
         } else { \
-            ucs_list_add_tail(&uct_ib_md_ops_list, &entry.list); \
+            ucs_list_for_each(p, &uct_ib_md_ops_list, list) { \
+                if (p->priority < _priority) { \
```
Comment (yosefe) [inline src/uct/ib/base/ib_md.h]: what if all priorities in the list are higher? it will not be added?

File: src/uct/ib/mlx5/dv/ib_mlx5dv_md.c [hunk 2]
```diff
@@ -25,29 +22,9 @@ typedef struct uct_ib_mlx5_dbrec_page {
...
-    return 0;
-#endif
-}
-
-static uct_ib_md_ops_t uct_ib_mlx5_md_ops;
-
 #if HAVE_DEVX
 
-static ucs_status_t uct_ib_mlx5dv_reg_atomic_key(uct_ib_md_t *ibmd,
+static ucs_status_t uct_ib_mlx5_devx_reg_atomic_key(uct_ib_md_t *ibmd,
```
Comment (yosefe) [inline src/uct/ib/mlx5/dv/ib_mlx5dv_md.c]: fix indentation in next line

File: src/uct/ib/mlx5/dv/ib_mlx5dv_md.c [hunk 3]
```diff
@@ -134,7 +111,7 @@ static ucs_status_t uct_ib_mlx5dv_reg_atomic_key(uct_ib_md_t *ibmd,
     return status;
 }
 
-static ucs_status_t uct_ib_mlx5dv_dereg_atomic_key(uct_ib_md_t *ibmd,
+static ucs_status_t uct_ib_mlx5_devx_dereg_atomic_key(uct_ib_md_t *ibmd,
```
Comment (yosefe) [inline src/uct/ib/mlx5/dv/ib_mlx5dv_md.c]: fix indentation in next line

File: src/uct/ib/mlx5/dv/ib_mlx5dv_md.c [hunk 4]
```diff
@@ -199,8 +176,7 @@ static ucs_mpool_ops_t uct_ib_mlx5_dbrec_ops = {
     .obj_cleanup   = NULL
 };
 
-/* assuming a DEVX cabable driver also implements ODP */
-static ucs_status_t uct_ib_mlx5dv_check_odp(uct_ib_mlx5_md_t *md, void *cap)
+static ucs_status_t uct_ib_mlx5_check_odp(uct_ib_mlx5_md_t *md, void *cap)
```
Comment (yosefe) [inline src/uct/ib/mlx5/dv/ib_mlx5dv_md.c]: uct_ib_mlx5_devx_check_odp

File: src/uct/ib/mlx5/dv/ib_mlx5dv_md.c [hunk 5]
```diff
@@ -257,7 +233,9 @@ static ucs_status_t uct_ib_mlx5dv_check_odp(uct_ib_mlx5_md_t *md, void *cap)
     return UCS_OK;
 }
 
-static ucs_status_t uct_ib_mlx5dv_md_open(struct ibv_device *ibv_device,
+static uct_ib_md_ops_t uct_ib_mlx5_devx_md_ops;
+
+static ucs_status_t uct_ib_mlx5_devx_md_open(struct ibv_device *ibv_device,
```
Comment (yosefe) [inline src/uct/ib/mlx5/dv/ib_mlx5dv_md.c]: fix indentation in next line

File: src/uct/ib/mlx5/dv/ib_mlx5dv_md.c [hunk 6]
```diff
@@ -407,10 +385,25 @@ static ucs_status_t uct_ib_mlx5dv_md_open(struct ibv_device *ibv_device,
     return status;
 }
 
-#else
+void uct_ib_mlx5_devx_md_cleanup(uct_ib_md_t *ibmd) {
```
Comment (yosefe) [inline src/uct/ib/mlx5/dv/ib_mlx5dv_md.c]: { on next line

File: src/uct/ib/mlx5/exp/ib_exp_md.c [hunk 7]
```diff
@@ -0,0 +1,461 @@
...
+#include <ucs/profile/profile.h>
+
+typedef struct uct_ib_mlx5_mem {
+    uct_ib_mem_t             super;
+#if HAVE_EXP_UMR
+    struct ibv_mr            *atomic_mr;
+#endif
+} uct_ib_mlx5_mem_t;
+
+static ucs_status_t uct_ib_mlx5_md_umr_qp_create(uct_ib_mlx5_md_t *md)
```
Comment (yosefe) [inline src/uct/ib/mlx5/exp/ib_exp_md.c]: uct_ib_mlx5_exp_md_umr_qp_create

File: src/uct/ib/mlx5/exp/ib_exp_md.c [hunk 8]
```diff
@@ -0,0 +1,461 @@
...
+err_destroy_cq:
+    ibv_destroy_cq(md->umr_cq);
+err:
+    return UCS_ERR_IO_ERROR;
+#else
+    return UCS_ERR_UNSUPPORTED;
+#endif
+}
+
+static ucs_status_t uct_ib_mlx5_verbs_reg_atomic_key(uct_ib_md_t *ibmd,
```
Comment (yosefe) [inline src/uct/ib/mlx5/exp/ib_exp_md.c]: uct_ib_mlx5_exp_md_reg_atomic_key

File: src/uct/ib/mlx5/exp/ib_exp_md.c [hunk 9]
```diff
@@ -0,0 +1,461 @@
...
+    UCS_PROFILE_CALL(ibv_dereg_mr, umr);
+err:
+    ucs_free(mem_reg);
+    return status;
+#else
+    return UCS_ERR_UNSUPPORTED;
+#endif
+}
+
+static ucs_status_t uct_ib_mlx5_verbs_dereg_atomic_key(uct_ib_md_t *ibmd,
```
Comment (yosefe) [inline src/uct/ib/mlx5/exp/ib_exp_md.c]: uct_ib_mlx5_exp_md_dereg_atomic_key

File: src/uct/ib/mlx5/exp/ib_exp_md.c [hunk 10]
```diff
@@ -0,0 +1,461 @@
...
+
+    return UCS_OK;
+#else
+    return UCS_ERR_UNSUPPORTED;
+#endif
+}
+
+static uct_ib_md_ops_t uct_ib_mlx5_md_ops;
+
+static ucs_status_t uct_ib_mlx5_verbs_md_open(struct ibv_device *ibv_device,
```
Comment (yosefe) [inline src/uct/ib/mlx5/exp/ib_exp_md.c]: uct_ib_mlx5_exp_md_open

File: src/uct/ib/mlx5/exp/ib_exp_md.c [hunk 11]
```diff
@@ -0,0 +1,461 @@
...
+
+err_free:
+    ucs_free(md);
+err_free_context:
+    ibv_close_device(ctx);
+err:
+    return status;
+}
+
+void uct_ib_mlx5_verbs_md_cleanup(uct_ib_md_t *ibmd) {
```
Comment (yosefe) [inline src/uct/ib/mlx5/exp/ib_exp_md.c]: uct_ib_mlx5_exp_md_cleanup
{ on next line

File: src/uct/ib/mlx5/ib_mlx5.h [hunk 12]
```diff
@@ -36,9 +36,9 @@
 #  include <infiniband/mlx5dv.h>
 #else
 #  include <infiniband/mlx5_hw.h>
-#  include "ib_mlx5_hw.h"
+#  include "exp/ib_mlx5_hw.h"
```
Comment (yosefe) [inline src/uct/ib/mlx5/ib_mlx5.h]: can we include with path. eg `<uct/ib/mlx5/exp/ib_mlx5_hw.h>`

File: src/uct/ib/mlx5/ib_mlx5.h [hunk 13]
```diff
@@ -36,9 +36,9 @@
...
 #endif
-#include "ib_mlx5_dv.h"
+#include "dv/ib_mlx5_dv.h"
```
Comment (yosefe) [inline src/uct/ib/mlx5/ib_mlx5.h]: don't we need some condition to include this? what happens on mofed?

Comment (Artemy-Mellanox) [reply src/uct/ib/mlx5/ib_mlx5.h]: No, uct_ib_mlx5dv_init_obj() API we provide unconditionally.
If MOFED doesn't provide underlying  mlx5dv_init_obj() we emulate it using mlx5_hw.h

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/10640/ for details (Mellanox internal link).

**Pull request merged**